pyro/contrib/gp/util.py

```diff
@@ -1,139 +1,237 @@
+    use :func:`torch.set_default_tensor_type` to change default tensor type). To cast
+    these parameters to a correct data type or GPU device, we can call methods such as
     :meth:`~torch.nn.Module.double` or :meth:`~torch.nn.Module.cuda`. See
     :class:`torch.nn.Module` for more information.
-
-    :param str name: Name of this object.
     """
-    def __init__(self, name=None):
+    def __init__(self):
         super(Parameterized, self).__init__()
-        self._priors = {}
-        self._constraints = {}
-        self._fixed_params = {}
-        self._registered_params = {}
+        self._constraints = OrderedDict()
+        self._priors = OrderedDict()
+        self._guides = OrderedDict()
+        self._mode = None
 
-        self.name = name
+    def __call__(self, *args, **kwargs):
+        with autoname.name_count():
+            return super(Parameterized, self).__call__(*args, **kwargs)
 
-    def set_prior(self, param, prior):
+    def set_constraint(self, name, constraint):
+        """
+        Sets a constraint to a parameter.
+
+        :param str name: Name of the parameter.
+        :param ~torch.distributions.constraints.Constraint constraint: A PyTorch
+            constraint. See :mod:`torch.distributions.constraints` for a list of
+            constraints.
+        """
+        if name in self._parameters:
+            # move param to _buffers
+            p = self._parameters.pop(name)
+            self.register_buffer(name, p)
+        elif name in self._buffers:
+            p = self._buffers[name]
+        else:
+            raise ValueError("There is no parameter with name: {}".format(name))
+
+        p_unconstrained = Parameter(transform_to(constraint).inv(p).detach())
+        self.register_parameter("{}_unconstrained".format(name), p_unconstrained)
+        self._constraints[name] = constraint
+
+    def set_prior(self, name, prior):
         """
         Sets a prior to a parameter.
 
-        :param str param: Name of the parameter.
+        :param str name: Name of the parameter.
         :param ~pyro.distributions.distribution.Distribution prior: A Pyro prior
             distribution.
         """
-        self._priors[param] = prior
+        if name in self._parameters:
+            # move param to _buffers
+            p = self._parameters.pop(name)
+            self.register_buffer(name, p)
+        elif name not in self._buffers:
+            raise ValueError("There is no parameter with name: {}".format(name))
 
-    def set_constraint(self, param, constraint):
-        """
-        Sets a constraint to a parameter.
+        self._priors[name] = prior
+        self.autoguide(name, dist.Delta)
 
-        :param str param: Name of the parameter.
-        :param ~torch.distributions.constraints.Constraint constraint: A PyTorch
-            constraint. See :mod:`torch.distributions.constraints` for a list of
-            constraints.
+    def autoguide(self, name, dist_constructor):
```

Comment:
It would be nice if we can use `contrib.autoguide` directly. However, there are a few conflicts here:
- `contrib.autoguide` will flatten the parameter then build a guide on it. Here we keep the shape of parameter as-is.
- `contrib.autoguide` works with ParamStoreDict so cann't  be trained with torch.optim.